### PR TITLE
Add /shame <user> command to put users to shame

### PR DESCRIPTION
**Pull request by kzfubar**

### CODE REVIEW

File: shame_script.py [hunk 1]
```diff
@@ -10,6 +10,7 @@
 from table2ascii import Alignment, TableStyle, table2ascii
 
 from discord_signup import signup
+from shame_command import shame
```
Comment (rocket3989) [inline shame_script.py]: imports should be sorted by category, then alpha. I think this is possible with ruff, but might need more cli args

File: shame_command.py [hunk 2]
```diff
@@ -0,0 +1,80 @@
...
+def get_todoist_token(email: str) -> str:
+    config.read(config_path)
+    if email in config["TODOIST_KEY_BY_EMAIL"]:
+        return config["TODOIST_KEY_BY_EMAIL"][email]
+    else:
+        return None
+
+
+# Get tasks from Todoist API with 'shame' label using aiohttp
+async def get_shame_tasks(todoist_token: str, session: aiohttp.ClientSession) -> list:
```
Comment (rocket3989) [inline shame_command.py]: I feel this is a good time to extract all the todoist logic to a separate method, sharing similar functions like this one?

also `list` is insufficient typing

File: shame_command.py [hunk 3]
```diff
@@ -0,0 +1,80 @@
...
+        return
+
+    async with aiohttp.ClientSession() as session:
+        # Get the tasks with the "shame" label
+        shame_tasks = await get_shame_tasks(todoist_token, session)
+
+        if not shame_tasks:
+            await interaction.followup.send(
+                f"{user_to_shame.mention} has nothing to be ashamed of!"
+            )
```
Comment (rocket3989) [inline shame_command.py:82]: nit: if this loop condition is met, there is nothing else that needs to be executed in the function, so it should `return`, and remove the else

File: shame_command.py [hunk 4]
```diff
@@ -0,0 +1,80 @@
...
+@commands.command(name="shame")
+async def shame(interaction: discord.Interaction, user_to_shame: discord.Member):
+    config.read(config_path)
+    email = None
+
+    # Find the email associated with the user
+    for stored_email, discord_id in config["DISCORD_ID_BY_EMAIL"].items():
+        if discord_id == str(user_to_shame.id):
+            email = stored_email
+            break
```
Comment (rocket3989) [inline shame_command.py]: There is a nice pattern to find the first predicate match:
https://stackoverflow.com/a/8534381

File: shame_script.py [hunk 5]
```diff
@@ -272,4 +273,23 @@ async def signup_passthrough(
...
+    await interaction.response.defer(ephemeral=False, thinking=True)
+    try:
+        await shame(interaction, user_to_shame)
+        logger.info("Shaming successful for user: %s", user_to_shame.name)
+
+    except Exception as e:
+        logger.exception("Error during shaming: %s", e)
+        await interaction.followup.send(
+            "An error occurred while processing the shame command."
+        )
```
Comment (rocket3989) [inline shame_script.py]: very nice, should be added to all slash commands

File: shame_command.py [hunk 6]
```diff
@@ -0,0 +1,80 @@
...
+
+# Get tasks from Todoist API with 'shame' label using aiohttp
+async def get_shame_tasks(todoist_token: str, session: aiohttp.ClientSession) -> list:
+    headers = {"Authorization": f"Bearer {todoist_token}"}
+    params = {"filter": "label:shame"}
+
+    async with session.get(TODOIST_API, headers=headers, params=params) as response:
+        if response.status != 200:
+            logger.error(f"Failed to retrieve tasks: {response.status}")
+            return []
```
Comment (rocket3989) [inline shame_command.py]: catching is good here, but this feels like a quiet failure mode for the end user- if the call fails it would just show that the user has no tasks

maybe an optional None return value that the caller can catch would work? or even just throwing an exception and catching it higher in the chain (if the task pull fails, the rest of the execution is meaningless)

**Review by rocket3989 — APPROVED** [review]
looks good, just some minor nits

I think the refactor can be a separate branch

File: shame_command.py [hunk 7]
```diff
@@ -0,0 +1,80 @@
...
+        else:
+            task_list = "\n".join([task["content"] for task in shame_tasks])
+            await interaction.followup.send(
+                f"For shame {user_to_shame.mention}!\n{task_list}"
+            )
+
+
+# Add this command to the bot
+def setup(bot):
+    bot.add_command(shame)
```
Comment (rocket3989) [inline shame_command.py]: Why is this here?

File: shame_command.py [hunk 8]
```diff
@@ -0,0 +1,80 @@
...
+
+
+# Get tasks from Todoist API with 'shame' label using aiohttp
+async def get_shame_tasks(todoist_token: str, session: aiohttp.ClientSession) -> list:
+    headers = {"Authorization": f"Bearer {todoist_token}"}
+    params = {"filter": "label:shame"}
+
+    async with session.get(TODOIST_API, headers=headers, params=params) as response:
+        if response.status != 200:
+            logger.error(f"Failed to retrieve tasks: {response.status}")
```
Comment (rocket3989) [inline shame_command.py]: log statements should not use f-strings

File: shame_script.py (hunk 5 above)
Comment (kzfubar) [reply shame_script.py]: maybe a followup change to add a base 'Command'

File: shame_command.py (hunk 2 above)
Comment (kzfubar) [reply shame_command.py]: Changed list -> List[Task]

sounds good to refactor as separate change

File: shame_command.py (hunk 3 above)
Comment (kzfubar) [reply shame_command.py:82]: Done

File: shame_command.py (hunk 4 above)
Comment (kzfubar) [reply shame_command.py]: I'm not entirely convinced that the pattern is more clear than writing the change out, but I don't feel too strongly one way or the other

File: shame_command.py (hunk 6 above)
Comment (kzfubar) [reply shame_command.py]: yeah I think this is a good point, since a client response error seems outside the normal operation, lets throw an exception here

File: shame_command.py (hunk 7 above)
Comment (kzfubar) [reply shame_command.py]: *you saw nothing*

File: shame_command.py (hunk 8 above)
Comment (kzfubar) [reply shame_command.py]: oh nice catch, thanks
